### ISSUE TITLE
Blur frozen on Android API 31

### ISSUE THREAD

**Issue by jakobulbrich** [chrisbanes/haze#77]
On Android 12 (API 31) the blur is frozen while scrolling.
Only when the bottom navigation appears or disappears the blur quickly updates and is then frozen again.

Seems like `HazeNode31.onObservedReadsChanged()` is not called. Probably a bug in Compose or can we do something about it?

See attached video of the android-jetpack sample on API 31. It is working as expected on API 32.

https://github.com/chrisbanes/haze/assets/16408565/efc8f36b-0637-4e3f-b132-13311aca464b

**Comment by chrisbanes** [comment]
Hmmm, might well be a platform issue of RenderNode in API 31 😢. Looking now.

**Comment by chrisbanes** [comment]
So it seems that the `draw()` function isn't being called consistently on API 31, but is on API 32+. I'll raise on the Compose Issue Tracker.

**Comment by chrisbanes** [comment]
https://issuetracker.google.com/issues/318679450

**Comment by chrisbanes** [comment]
I'll have another play tomorrow and see if I can get something working. If not I'll probably have to bump up the min SDK level for the blur impl to 32 for now.

**Comment by gleb-skobinsky** [comment]
I hope I can add some context as to the origins and the reason why this bug happens. It appears that everything is okay on the Compose side. The issue here is with the RenderNode. My hypothesis is that on API 31, it fails to refresh on the JNI side if its position has not updated since the latest draw call. More specifically, in this part of the code:
```
private fun Effect.updateRenderNodePosition() {
    renderNode.setPosition(0, 0, bounds.width.toInt(), bounds.height.toInt())
    renderNode.translationX = bounds.left
    renderNode.translationY = bounds.top
    renderNodeDirty = false
  }
  ```
  If we feed different positions at each call, the RenderNode is updated successfully. It is not acceptable to use in the library, of course, but I wrote a little dirty hack to reproduce this behavour: https://github.com/gleb-skobinsky/haze/blob/main/haze/src/androidMain/kotlin/dev/chrisbanes/haze/AndroidHazeNode.kt
  
https://github.com/chrisbanes/haze/assets/90451272/7853ba32-dc82-471d-99f7-3ce7c6020a79

**Comment by gleb-skobinsky** [comment]
The video is so short because github does not accept long screen recordings :)

**Comment by gleb-skobinsky** [comment]
UPD: I managed to remove the freeze on SDK level 31 without the hacks! Just applied the translation to the `setPosition` method instead of translation in the `record` lambda. Check out the <a href="https://github.com/gleb-skobinsky/haze/blob/main/haze/src/androidMain/kotlin/dev/chrisbanes/haze/AndroidHazeNode.kt">source code</a>

**Comment by gleb-skobinsky** [comment]
It took hours to debug this...

**Comment by chrisbanes** [comment]
Great detective work @gleb-skobinsky! Send a PR over?

**Comment by gleb-skobinsky** [comment]
Yes, I will create a PR today

**Comment by chrisbanes** [comment]
Going to close this, as there's not much we can do. RenderNode received a number of improvements and fixes in Android SDK 32 which we rely on. Therefore we'll need to keep our current behaviour of not using the blur effect on SDK 31.

**Comment by Skaldebane** [comment]
Could this be an emulator-specific quirk? I do encounter this bug on an Android 12 emulator, but no physical device I tried thus far exerts this behavior.

I have a Samsung Galaxy A11 running Android 12, and it works fine there.
I also tried some Firebase Remote devices with SDK 31:

    Samsung Galaxy A51: Haze works properly, content is updated properly.
    Google Pixel 6a: Haze works properly. And this is the closest we can be to AOSP.

There's definitely better ways to get access to more devices, but I guess this paints a good general picture of what's going on. I suspect the emulated graphics in the emulator have something to do with it, but I may be wrong about that.

Though I'd love to see a Xiaomi being tested, since they're notorious for creating their own bugs (let alone fixing AOSP ones), and any other devices.

**Comment by chrisbanes** [comment]
Just like @Skaldebane, I just tried the Haze Samples on a Pixel 6 and Samsung A51 on **SDK Level 31** and everything seems to be running fine.

I see two paths forward:

1. Enable blurring on SDK Level 31 (also known as the _yolo_ plan).
2. Add a flag which allows devs to `enableBlurringOnProblematicPlatforms`, probably defaulting to `false`. When true, we enable the effect on SDK 31.

I'm leaning to option 2, but happy to hear suggestions.

**Comment by chrisbanes** [comment]
Thinking about it, I think a `blurEnabled: () -> Boolean` block is the way to go.

It will default to calling a public function, but you can override it however you wish. This is platform-agnostic too, which is a big bonus.

**Comment by chrisbanes** [comment]
> I assume it's just going to ignore the value on SDK lower than 31 on Android, right?

Yes, exactly. It will ignore values where blurring is not feasible.

**Comment by chrisbanes** [comment]
One the PR above lands, you'll be able to do the following and enable blurring everywhere (where it's technically possible):

``` kotlin
Modifier.hazeChild(...) {
  blurEnabled = true
}
```

**Comment by krzysztof-lopacki-gft** [comment]
I confirm the "frozen blur issue" on Oppo Reno 6, Android 12 (API 31). 
It proves not only emulators are affected.

**Comment by chrisbanes** [comment]
Just circling back here. I've found a fix for this and it will be released in 1.6.0 (next week).

I will have already enabled API 31 support by default, so there's nothing to do here.
